scrapy/dupefilters.py

```diff
@@ -24,35 +26,49 @@ def log(self, request, spider):  # log that a request has been filtered
         pass
 
 
+def _escape_line_breaks(data):
+    """Returns `data` with escaped line breaks"""
+    return data.replace(b'\\', b'\\\\').replace(b'\n', b'\\n')
+
+
+def _unescape_line_breaks(data):
+    """Performs the reverse process of
+    :func:`scrapy.dupefilters._escape_line_breaks`."""
+    value = re.sub(b'(^|[^\\\\])((?:\\\\\\\\)*?)\\\\n', b'\\1\\2\n', data)
+    return value.replace(b'\\\\', b'\\')
+
+
 class RFPDupeFilter(BaseDupeFilter):
     """Request Fingerprint duplicates filter"""
 
-    def __init__(self, path=None, debug=False):
+    def __init__(self, path=None, debug=False, *,
+                 key_builder=default_request_key_builder):
         self.file = None
-        self.fingerprints = set()
+        self.build_key = key_builder
+        self.keys = set()
         self.logdupes = True
         self.debug = debug
         self.logger = logging.getLogger(__name__)
         if path:
-            self.file = open(os.path.join(path, 'requests.seen'), 'a+')
+            request_seen_path = os.path.join(path, 'requests.seen')
+            self.file = open(request_seen_path, 'ab+', newline='\n')
             self.file.seek(0)
-            self.fingerprints.update(x.rstrip() for x in self.file)
+            for escaped_key in self.file:
+                self.keys.add(_unescape_line_breaks(escaped_key[:-1]))
 
     @classmethod
     def from_settings(cls, settings):
         debug = settings.getbool('DUPEFILTER_DEBUG')
-        return cls(job_dir(settings), debug)
+        key_builder = load_object(settings['REQUEST_KEY_BUILDER'])
+        return cls(job_dir(settings), debug, key_builder=key_builder)
 
     def request_seen(self, request):
-        fp = self.request_fingerprint(request)
-        if fp in self.fingerprints:
+        key = self.build_key(request)
+        if key in self.keys:
```

Comment:
The idea behind the general name change, from fingerprint to key, was to imply that the resulting values should not collide, otherwise there would be undefined behavior. Fingerprints can collide (although it should be extremely unlikely). Keys cannot collide.

`request_key` would work here, though.